### PR TITLE
[scripts] Download arm64 python when host is arm64 on windows.

### LINKED ISSUE #34353: [dav1d] build fails on windows on arm
### Operating system

Windows on ARM

### Compiler

MSVC

### Steps to reproduce the behavior

```Shell
./vcpkg.exe install dav1d --triplet arm64-windows-static
```

### Failure logs

```
...
[95/99] "C:/Users/lemourin/source/repos/vcpkg/downloads/tools/nasm/nasm-2.16.01\nasm.EXE" "-f" "win64" "-I" "C:/Users/lemourin/source/repos/vcpkg/buildtrees/dav1d/src/1.2.1-f691f696b3/src/" "-I" "C:/Users/lemourin/source/repos/vcpkg/buildtrees/dav1d/arm64-windows-static-dbg/" "-MQ" "src/libdav1d.a.p/itx16_sse.obj" "-MF" "src/libdav1d.a.p/itx16_sse.obj.ndep" "../src/1.2.1-f691f696b3/src/x86/itx16_sse.asm" "-o" "src/libdav1d.a.p/itx16_sse.obj"
[96/99] "C:/Users/lemourin/source/repos/vcpkg/downloads/tools/nasm/nasm-2.16.01\nasm.EXE" "-f" "win64" "-I" "C:/Users/lemourin/source/repos/vcpkg/buildtrees/dav1d/src/1.2.1-f691f696b3/src/" "-I" "C:/Users/lemourin/source/repos/vcpkg/buildtrees/dav1d/arm64-windows-static-dbg/" "-MQ" "src/libdav1d.a.p/itx16_avx2.obj" "-MF" "src/libdav1d.a.p/itx16_avx2.obj.ndep" "../src/1.2.1-f691f696b3/src/x86/itx16_avx2.asm" "-o" "src/libdav1d.a.p/itx16_avx2.obj"
[97/99] "C:/Users/lemourin/source/repos/vcpkg/downloads/tools/nasm/nasm-2.16.01\nasm.EXE" "-f" "win64" "-I" "C:/Users/lemourin/source/repos/vcpkg/buildtrees/dav1d/src/1.2.1-f691f696b3/src/" "-I" "C:/Users/lemourin/source/repos/vcpkg/buildtrees/dav1d/arm64-windows-static-dbg/" "-MQ" "src/libdav1d.a.p/mc16_sse.obj" "-MF" "src/libdav1d.a.p/mc16_sse.obj.ndep" "../src/1.2.1-f691f696b3/src/x86/mc16_sse.asm" "-o" "src/libdav1d.a.p/mc16_sse.obj"
[98/99] "C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.37.32822/bin/Hostarm64/arm64/lib.exe" "-machine:ARM64" "-nologo" "/NOLOGO" /MACHINE:ARM64 /OUT:src/libdav1d.a @src/libdav1d.a.rsp
FAILED: src/libdav1d.a
"C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.37.32822/bin/Hostarm64/arm64/lib.exe" "-machine:ARM64" "-nologo" "/NOLOGO" /MACHINE:ARM64 /OUT:src/libdav1d.a @src/libdav1d.a.rsp
src\libdav1d.a.p\cpuid.obj : fatal error LNK1112: module machine type 'x64' conflicts with target machine type 'ARM64'
```

### Additional context

It builds when I substitute python that vcpkg is using with python built for windows for arm. This is related to https://github.com/microsoft/vcpkg/discussions/25444.

### PR DESCRIPTION
**Pull request by lemourin**
Attempt to fix [1]. It should download arm64 python when host is arm64, not the target.

[1]: https://github.com/microsoft/vcpkg/issues/34353

### CODE REVIEW

**Comment by dg0yt** [comment]
x86 windows might be a little bit special but ... in general, these scripts must check the host system architecture, not the target system architecture. Something like:
https://github.com/microsoft/vcpkg/blob/cc97b4536ae749ec0e4f643488b600b217540fb3/scripts/cmake/vcpkg_configure_meson.cmake#L114-L137

**Comment by lemourin** [comment]
Thanks! What meson download script does seems to work for me.

**Comment by Osyotr** [comment]
Related: https://github.com/microsoft/vcpkg/pull/31727/commits/a4c1a7c61821a49fe7086390ea095311e7845367

**Comment by dg0yt** [comment]
> Related: [a4c1a7c](https://github.com/microsoft/vcpkg/commit/a4c1a7c61821a49fe7086390ea095311e7845367) (from #31727)

That commit doesn't much about its background. And gobject-introspection may still have its own problems, cf. https://github.com/microsoft/vcpkg/pull/33792#issuecomment-1733590043.

**Comment by jimwang118** [comment]
Thanks for your PR, Is work still being done for this PR?

**Comment by lemourin** [comment]
Not sure how to get around `gobject-introspection` failure:

```"D:\downloads\tools\python\python-3.11.5-x64\python.exe" "D:\installed\x64-windows\tools\meson\meson.py" "--internal" "exe" "--unpickle" "D:\buildtrees\gobject-introspection\x86-windows-rel\meson-private\meson_exe_python.exe_0c00ea554ecf45624a129e8881d00ea59433eb9b.dat"
while executing ['D:/downloads/tools/python/python-3.11.5-x64/python.exe', 'D:/buildtrees/gobject-introspection/x86-windows-rel/tools/g-ir-scanner', '--output=gir/GLib-2.0.gir', '--no-libtool', '--quiet', '--reparse-validate', '--add-include-path', 'D:/buildtrees/gobject-introspection/x86-windows-rel/gir', '--add-include-path', 'D:/buildtrees/gobject-introspection/src/spection-1-b8658e5018.clean/gir', '--identifier-prefix=G', '--symbol-prefix=g', '--symbol-prefix=glib', '--c-include=glib.h', '--namespace=GLib', '--nsversion=2.0', '--library=glib-2.0', '--library=gobject-2.0', '--external-library', '--pkg=glib-2.0', '--cflags-begin', '-ID:/installed/x86-windows/lib/pkgconfig//../../include/glib-2.0', '-ID:/installed/x86-windows/lib/pkgconfig//../../lib/glib-2.0/include', '-utf-8', '-DGLIB_COMPILATION', '-D__G_I18N_LIB_H__', '-DGETTEXT_PACKAGE=Dummy', '--cflags-end', 'D:/installed/x86-windows/lib/pkgconfig//../../include/glib-2.0/gobject/glib-types.h', 'D:/installed/x86-windows/lib/pkgconfig//../../lib/glib-2.0/include/glibconfig.h', 'D:/installed/x86-windows/lib/pkgconfig//../../include/glib-2.0/glib/galloca.h', 'D:/installed/x86-windows/lib/pkgconfig//../../include/glib-2.0/glib/garray.h', 'D:/installed/x86-windows/lib/pkgconfig//../../include/glib-2.0/glib/gasyncqueue.h', 'D:/installed/x86-windows/lib/pkgconfig//../../include/glib-2.0/glib/gatomic.h', 'D:/installed/x86-windows/lib/pkgconfig//../../include/glib-2.0/glib/gbacktrace.h', 'D:/installed/x86-windows/lib/pkgconfig//../../include/glib-2.0/glib/gbase64.h', 'D:/installed/x86-windows/lib/pkgconfig//../../include/glib-2.0/glib/gbitlock.h', 'D:/installed/x86-windows/lib/pkgconfig//../../include/glib-2.0/glib/gbookmarkfile.h', 'D:/installed/x86-windows/lib/pkgconfig//../../include/glib-2.0/glib/gbytes.h', 'D:/installed/x86-windows/lib/pkgconfig//../../include/glib-2.0/glib/gcharset.h', 'D:/installed/x86-windows/lib/pkgconfig//../../include/glib-2.0/glib/gchecksum.h', 'D:/installed/x86-windows/lib/pkgconfig//../../include/glib-2.0/glib/gconvert.h', 'D:/installed/x86-windows/lib/pkgconfig//../../include/glib-2.0/glib/gdataset.h', 'D:/installed/x86-windows/lib/pkgconfig//../../include/glib-2.0/glib/gdate.h', 'D:/installed/x86-windows/lib/pkgconfig//../../include/glib-2.0/glib/gdatetime.h', 'D:/installed/x86-windows/lib/pkgconfig//../../include/glib-2.0/glib/gdir.h', 'D:/installed/x86-windows/lib/pkgconfig//../../include/glib-2.0/glib/genviron.h', 'D:/installed/x86-windows/lib/pkgconfig//../../include/glib-2.0/glib/gerror.h', 'D:/installed/x86-windows/lib/pkgconfig//../../include/glib-2.0/glib/gfileutils.h', 'D:/installed/x86-windows/lib/pkgconfig//../../include/glib-2.0/glib/ggettext.h', 'D:/installed/x86-windows/lib/pkgconfig//../../include/glib-2.0/glib/ghash.h', 'D:/installed/x86-windows/lib/pkgconfig//../../include/glib-2.0/glib/ghmac.h', 'D:/installed/x86-windows/lib/pkgconfig//../../include/glib-2.0/glib/ghook.h', 'D:/installed/x86-windows/lib/pkgconfig//../../include/glib-2.0/glib/ghostutils.h', 'D:/installed/x86-windows/lib/pkgconfig//../../include/glib-2.0/glib/gi18n-lib.h', 'D:/installed/x86-windows/lib/pkgconfig//../../include/glib-2.0/glib/gi18n.h', 'D:/installed/x86-windows/lib/pkgconfig//../../include/glib-2.0/glib/giochannel.h', 'D:/installed/x86-windows/lib/pkgconfig//../../include/glib-2.0/glib/gkeyfile.h', 'D:/installed/x86-windows/lib/pkgconfig//../../include/glib-2.0/glib/glib-typeof.h', 'D:/installed/x86-windows/lib/pkgconfig//../../include/glib-2.0/glib/glib-visibility.h', 'D:/installed/x86-windows/lib/pkgconfig//../../include/glib-2.0/glib/glist.h', 'D:/installed/x86-windows/lib/pkgconfig//../../include/glib-2.0/glib/gmacros.h', 'D:/installed/x86-windows/lib/pkgconfig//../../include/glib-2.0/glib/gmain.h', 'D:/installed/x86-windows/lib/pkgconfig//../../include/glib-2.0/glib/gmappedfile.h', 'D:/installed/x86-windows/lib/pkgconfig//../../include/glib-2.0/glib/gmarkup.h', 'D:/installed/x86-windows/lib/pkgconfig//../../include/glib-2.0/glib/gmem.h', 'D:/installed/x86-windows/lib/pkgconfig//../../include/glib-2.0/glib/gmessages.h', 'D:/installed/x86-windows/lib/pkgconfig//../../include/glib-2.0/glib/gnode.h', 'D:/installed/x86-windows/lib/pkgconfig//../../include/glib-2.0/glib/goption.h', 'D:/installed/x86-windows/lib/pkgconfig//../../include/glib-2.0/glib/gpathbuf.h', 'D:/installed/x86-windows/lib/pkgconfig//../../include/glib-2.0/glib/gpattern.h', 'D:/installed/x86-windows/lib/pkgconfig//../../include/glib-2.0/glib/gpoll.h', 'D:/installed/x86-windows/lib/pkgconfig//../../include/glib-2.0/glib/gprimes.h', 'D:/installed/x86-windows/lib/pkgconfig//../../include/glib-2.0/glib/gprintf.h', 'D:/installed/x86-windows/lib/pkgconfig//../../include/glib-2.0/glib/gqsort.h', 'D:/installed/x86-windows/lib/pkgconfig//../../include/glib-2.0/glib/gquark.h', 'D:/installed/x86-windows/lib/pkgconfig//../../include/glib-2.0/glib/gqueue.h', 'D:/installed/x86-windows/lib/pkgconfig//../../include/glib-2.0/glib/grand.h', 'D:/installed/x86-windows/lib/pkgconfig//../../include/glib-2.0/glib/grcbox.h', 'D:/installed/x86-windows/lib/pkgconfig//../../include/glib-2.0/glib/grefcount.h', 'D:/installed/x86-windows/lib/pkgconfig//../../include/glib-2.0/glib/grefstring.h', 'D:/installed/x86-windows/lib/pkgconfig//../../include/glib-2.0/glib/gregex.h', 'D:/installed/x86-windows/lib/pkgconfig//../../include/glib-2.0/glib/gscanner.h', 'D:/installed/x86-windows/lib/pkgconfig//../../include/glib-2.0/glib/gsequence.h', 'D:/installed/x86-windows/lib/pkgconfig//../../include/glib-2.0/glib/gshell.h', 'D:/installed/x86-windows/lib/pkgconfig//../../include/glib-2.0/glib/gslice.h', 'D:/installed/x86-windows/lib/pkgconfig//../../include/glib-2.0/glib/gslist.h', 'D:/installed/x86-windows/lib/pkgconfig//../../include/glib-2.0/glib/gspawn.h', 'D:/installed/x86-windows/lib/pkgconfig//../../include/glib-2.0/glib/gstdio.h', 'D:/installed/x86-windows/lib/pkgconfig//../../include/glib-2.0/glib/gstrfuncs.h', 'D:/installed/x86-windows/lib/pkgconfig//../../include/glib-2.0/glib/gstring.h', 'D:/installed/x86-windows/lib/pkgconfig//../../include/glib-2.0/glib/gstringchunk.h', 'D:/installed/x86-windows/lib/pkgconfig//../../include/glib-2.0/glib/gstrvbuilder.h', 'D:/installed/x86-windows/lib/pkgconfig//../../include/glib-2.0/glib/gtestutils.h', 'D:/installed/x86-windows/lib/pkgconfig//../../include/glib-2.0/glib/gthread.h', 'D:/installed/x86-windows/lib/pkgconfig//../../include/glib-2.0/glib/gthreadpool.h', 'D:/installed/x86-windows/lib/pkgconfig//../../include/glib-2.0/glib/gtimer.h', 'D:/installed/x86-windows/lib/pkgconfig//../../include/glib-2.0/glib/gtimezone.h', 'D:/installed/x86-windows/lib/pkgconfig//../../include/glib-2.0/glib/gtrashstack.h', 'D:/installed/x86-windows/lib/pkgconfig//../../include/glib-2.0/glib/gtree.h', 'D:/installed/x86-windows/lib/pkgconfig//../../include/glib-2.0/glib/gtypes.h', 'D:/installed/x86-windows/lib/pkgconfig//../../include/glib-2.0/glib/gunicode.h', 'D:/installed/x86-windows/lib/pkgconfig//../../include/glib-2.0/glib/guri.h', 'D:/installed/x86-windows/lib/pkgconfig//../../include/glib-2.0/glib/gutils.h', 'D:/installed/x86-windows/lib/pkgconfig//../../include/glib-2.0/glib/guuid.h', 'D:/installed/x86-windows/lib/pkgconfig//../../include/glib-2.0/glib/gvariant.h', 'D:/installed/x86-windows/lib/pkgconfig//../../include/glib-2.0/glib/gvarianttype.h', 'D:/installed/x86-windows/lib/pkgconfig//../../include/glib-2.0/glib/gversion.h', 'D:/installed/x86-windows/lib/pkgconfig//../../include/glib-2.0/glib/gversionmacros.h', 'D:/installed/x86-windows/lib/pkgconfig//../../include/glib-2.0/glib/gwin32.h', '../src/spection-1-b8658e5018.clean/gir/glib-2.0.c']
--- stdout ---

--- stderr ---
Traceback (most recent call last):

  File "D:\buildtrees\gobject-introspection\x86-windows-rel\tools\g-ir-scanner", line 103, in <module>

    from giscanner.scannermain import scanner_main

  File "D:\buildtrees\gobject-introspection\x86-windows-rel\giscanner\scannermain.py", line 35, in <module>

    from giscanner.ast import Include, Namespace

  File "D:\buildtrees\gobject-introspection\x86-windows-rel\giscanner\ast.py", line 29, in <module>

    from .sourcescanner import CTYPE_TYPEDEF, CSYMBOL_TYPE_TYPEDEF

  File "D:\buildtrees\gobject-introspection\x86-windows-rel\giscanner\sourcescanner.py", line 34, in <module>

    from giscanner._giscanner import SourceScanner as CSourceScanner


ImportError: DLL load failed while importing _giscanner: %1 is not a valid Win32 application.
```

It seems to require python for target arch, which sounds wrong. I removed explicit win32 & x86 cross build support for it.

**Comment by dg0yt** [comment]
Port `gobject-introspection` has cross-build issues.

File: scripts/cmake/vcpkg_find_acquire_program(PYTHON3).cmake [hunk 1]
```diff
@@ -1,7 +1,17 @@
...
+        set(build_arch $ENV{PROCESSOR_ARCHITEW6432})
+    else()
+        set(build_arch $ENV{PROCESSOR_ARCHITECTURE})
+    endif()
+    if(build_arch MATCHES "^(ARM|arm)64$")
+        set(tool_subdirectory "python-${program_version}-arm64")
+        set(download_urls "https://www.python.org/ftp/python/${program_version}/python-${program_version}-embed-arm64.zip")
+        set(download_filename "python-${program_version}-embed-arm64.zip")
+        set(download_sha512 23d67fa89176fb47f564df9713039e86d1957831866c20d474be03fe6616bd15292e9912e44d3c1830747c24eb2131aeda53f19b3d4e7d1c8bd2bbb461e56f6c)
+    elseif(build_arch MATCHES "(x|X)86")
```
Comment (BillyONeal) [inline scripts/cmake/vcpkg_find_acquire_program(PYTHON3).cmake:14]:
```suggestion
    else()
```

Comment (BillyONeal) [reply scripts/cmake/vcpkg_find_acquire_program(PYTHON3).cmake:14]: I think we should 'give up and at least try' the x86 copy regardless of `build_arch`.

**Review by BillyONeal — CHANGES_REQUESTED** [review]

File: scripts/cmake/vcpkg_find_acquire_program(PYTHON3).cmake (hunk 1 above)
Comment (lemourin) [reply scripts/cmake/vcpkg_find_acquire_program(PYTHON3).cmake]: I changed the order of conditions, earlier I used x64 as 'catch all'.

**Comment by lemourin** [comment]
@data-queue Why was this converted to a draft? The CI failure is unrelated.

**Comment by lemourin** [comment]
Ping.

**Comment by jimwang118** [comment]
@BillyONeal Is there anything else that needs to be modified in this PR now?

**Review by BillyONeal — APPROVED** [review]

**Pull request merged**